ieproxy_windows.go

```diff
@@ -24,8 +25,86 @@ func getConf() ProxyConf {
 }
 
 func writeConf() {
-	regedit, _ := readRegedit()
-	windowsProxyConf = parseRegedit(regedit)
+	if cfg, err := getConfFromCall(); err == nil {
+		windowsProxyConf = ProxyConf{
+			Static: StaticProxyConf{
+				Active: cfg.lpszProxy != nil,
+			},
+			Script: ProxyScriptConf{
+				Active:       cfg.lpszAutoConfigUrl != nil || cfg.fAutoDetect,
+				AutoDiscover: cfg.fAutoDetect,
+			},
+		}
+
+		if windowsProxyConf.Static.Active {
+			protocol := make(map[string]string)
+			for _, s := range strings.Split(StringFromUTF16Ptr(cfg.lpszProxy), ";") {
+				if s == "" {
+					continue
+				}
+				pair := strings.SplitN(s, "=", 2)
+				if len(pair) > 1 {
+					protocol[pair[0]] = pair[1]
+				} else {
+					protocol[""] = pair[0]
+				}
+			}
+
+			windowsProxyConf.Static.Protocols = protocol
+			if cfg.lpszProxyBypass != nil {
+				windowsProxyConf.Static.NoProxy = strings.Replace(StringFromUTF16Ptr(cfg.lpszProxyBypass), ";", ",", -1)
+			}
+		}
+
+		if windowsProxyConf.Script.Active {
+			windowsProxyConf.Script.URL = StringFromUTF16Ptr(cfg.lpszAutoConfigUrl)
+		}
+	} else {
+		regedit, _ := readRegedit() //If the syscall fails, backup to manual detection.
+		windowsProxyConf = parseRegedit(regedit)
+	}
+}
+
+func getConfFromCall() (*tWINHTTP_CURRENT_USER_IE_PROXY_CONFIG, error) {
+	winHttp := syscall.NewLazyDLL("Winhttp.dll")
+	open := winHttp.NewProc("WinHttpOpen")
+	handle, _, err := open.Call(0, 0, 0, 0, 0)
+	if handle == 0 {
+		return &tWINHTTP_CURRENT_USER_IE_PROXY_CONFIG{}, err
+	}
+	close := winHttp.NewProc("WinHttpCloseHandle")
+	defer close.Call(handle)
+
+	getIEProxyConfigForCurrentUser := winHttp.NewProc("WinHttpGetIEProxyConfigForCurrentUser")
+
+	config := new(tWINHTTP_CURRENT_USER_IE_PROXY_CONFIG)
+
+	ret, _, err := getIEProxyConfigForCurrentUser.Call(uintptr(unsafe.Pointer(config)))
+	if ret > 0 {
+		err = nil
+	}
+
+	if config.fAutoDetect == true {
+		detectAutoProxyConfigURL := winHttp.NewProc("WinHttpDetectAutoProxyConfigUrl")
+		adFlag := uintptr(fWINHTTP_AUTO_DETECT_TYPE_DNS_A | fWINHTTP_AUTO_DETECT_TYPE_DHCP)
+		acURL := new(uint16)
+
+		ret, _, err = detectAutoProxyConfigURL.Call(
+			adFlag,
+			uintptr(unsafe.Pointer(acURL)),
+		)
+		if ret > 0 {
+			err = nil
+		}
+
+		//if err.(syscall.Errno) == syscall.Errno(12180) {
+		if err != nil {
+			config.fAutoDetect = false
+			err = nil
+		}
+	}
+
+	return config, err
```

Comment:
As per the docs for WinHttpGetIEProxyConfigForCurrentUser, lpszProxy, lpszProxyBypass and lpszAutoConfigUrl all need to be freed by us, using globalFree, if they are not nil.  See comment on pac_windows for details, and suggested impl.

Note that this fact also implies that our return value from this function should not be a Windows struct (since then caller would need to do the freeing, which gets ugly. Better to copy the stuff we want to return out of the  *tWINHTTP_CURRENT_USER_IE_PROXY_CONFIG right here, into string-type fields our own go-ish struct that we define (i.e. calls to StringFromUTF16Ptr are here) , and then, after we've made those copies, free the lpsz*s